alf/utils/distributions.py

```diff
@@ -0,0 +1,318 @@
+
+    @staticmethod
+    def log_prob(x):
+        # log(sqrt(2 * pi)) = 0.9189385332046727
+        return -0.5 * (x**2) - 0.9189385332046727
+
+
+class CauchyITS(InverseTransformSampling):
+    """Cauchy distribution.
+
+    .. math::
+
+        p(x) = 1 / (pi * (1 + x*x))
+
+    """
+
+    @staticmethod
+    def cdf(x):
+        return torch.atan(x) / math.pi + 0.5
+
+    @staticmethod
+    def icdf(x):
+        return torch.tan(math.pi * (x - 0.5))
+
+    @staticmethod
+    def log_prob(x):
+        return -(math.pi * (1 + x**2)).log()
+
+
+class T2Cdf_(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x):
+        # return 0.5 + 0.5 * x / (1 + x**2).sqrt()
+        # The original form is not numerically increasing (i.e. for some large
+        # value of x2>x1, f(x2) < f(x1) because of insufficient numerical
+        # precision), so we use the following form. And we need to implement our
+        # own backward because autograd cannot calculate the gradient of the
+        # following form when x is 0.
+        y = 0.5 + 0.5 * x.sign() * (1 + x**-2).rsqrt()
+        ctx.save_for_backward(x)
+        return y
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        x, = ctx.saved_tensors
+        return 0.5 * (1 + x**2)**(-1.5) * grad_output
+
+
+t2cdf = T2Cdf_.apply
+
+
+class T2ITS(InverseTransformSampling):
+    """Student's t-distribution with DOF 2.
+
+    .. math::
+
+        p(x) = 1 / (2 * (1 + x*x) ** 1.5)
+
+    """
+
+    @staticmethod
+    def cdf(x):
+        return t2cdf(x)
+
+    @staticmethod
+    def icdf(x):
+        return (x - 0.5) * (x * (1 - x)).rsqrt()
+
+    @staticmethod
+    def log_prob(x):
+        # log(2) = 0.6931471805599453
+        return -1.5 * (1 + x**2).log() - 0.6931471805599453
+
+
+class TruncatedDistribution(td.Distribution):
+    r"""The base class of truncated distributions.
+
+    A truncated distribution :math:`q(x)` is defined as a standard base distribution :math:`p(x)` and
+    location :math:`\mu`, scale parameters :math:`s`, lower bound :math:`l` and
+    upper bound :math:`u`
+
+    .. math::
+
+        q(x) = \frac{1}{s (P(u) - P(l))}p(\frac{x-\mu}{s})
+
+    where :math:`P` is the cdf of :math:`p`.
+
+    Args:
+        loc: the location parameter
+        scale: the scale parameter
+        lower_bound: the lower bound
+        upper_bound: the upper bound
+        its: the standard distribution to be used.
+    """
+
+    arg_constraints = {
+        'loc': td.constraints.real,
+        'scale': td.constraints.positive
+    }
+    has_rsample = True
+
+    def __init__(self, loc: Tensor, scale: Tensor, lower_bound: Tensor,
+                 upper_bound: Tensor, its: InverseTransformSampling):
+        event_shape = torch.broadcast_shapes(lower_bound.shape,
+                                             upper_bound.shape)
+        batch_shape = torch.broadcast_shapes(scale.shape, loc.shape,
+                                             event_shape)
```

Comment:
This seems to imply `scale` and `loc`, unlike `lower_bound` and `upper_bound`, may have batch dimension. If so, should we explicitly state this in the docstring?

Also it is not entirely clear that `lower_bound` and `upper_bound` ('s broadcast) defines the shape of the event from the docstring.